From: SeongJae Park <sj@kernel.org>
Date: Wed, 3 Sep 2025 18:19:17 -0700
Subject: [PATCH] mm/damon/core: reduce total_charged_{sz,ns} to avoid
 overflows

When a scheme having a time quota is applied, and if total_charged_sz or
total_charged_ns are overflowed, the throughput that calculated using
the two counters can be unexpectedly small or high, respectively.  As a
result, users can see the scheme working unexpectedly slow or fast.  It
is unlikely in 64 bit, but can happen in 32 bit.  Avoid the problem by
reducing the counters to use only up to half of their max value.

Signed-off-by: SeongJae Park <sj@kernel.org>
---
 mm/damon/core.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/mm/damon/core.c b/mm/damon/core.c
index 6cb29457bd32..76cd0666b93d 100644
--- a/mm/damon/core.c
+++ b/mm/damon/core.c
@@ -2305,7 +2305,8 @@ static void damos_adjust_quota(struct damon_ctx *c, struct damos *s)
 		if (quota->esz && quota->charged_sz >= quota->esz)
 			s->stat.qt_exceeds++;
 		quota->total_charged_sz += quota->charged_sz;
-		if (quota->total_charged_sz > ULONG_MAX / 2) {
+		if (quota->total_charged_sz > ULONG_MAX / 2 ||
+				quota->total_charged_ns > ULONG_MAX / 2) {
 			quota->total_charged_sz /= 2;
 			quota->total_charged_ns /= 2;
 		}
-- 
2.39.5

